File: examples/ros2/ros2_ws/src/sheaf_ros2/sheaf_ros2/protocol.py
"""Python side of the sheaf/ROS 2 wire protocol (v2).

Mirrors examples/ros2/protocol.jl EXACTLY. Change the message format here AND in
protocol.jl together -- nowhere else.

Frames: everything is ENU, in the mocap / Gazebo world frame. Metres, rad, s.

This module is deliberately free of any ROS or numpy import so it can be unit
tested (and reused by scripts/fake_sim.py) without a ROS 2 installation.
"""

from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, Dict, List, Sequence

PROTOCOL_VERSION = 2


@dataclass
class AgentState:
    id: int
    kind: str  # "px4" | "diffdrive"
    p: Sequence[float]  # position [x, y, z]
    v: Sequence[float]  # velocity [vx, vy, vz]
    yaw: float
    ok: bool  # False => state is stale/invalid, controller should hold this agent


@dataclass
class AgentControl:
    id: int
    u: List[float]  # commanded world-frame ENU velocity
    yaw_rate: float


@dataclass
class TargetPose:
    id: int
    p: List[float]


@dataclass
class ControlOutput:
    agents: List[AgentControl]
    targets: List[TargetPose] = field(default_factory=list)
    diag: Dict[str, Any] = field(default_factory=dict)


def format_state(t: float, agents: Sequence[AgentState], stop: bool = False) -> bytes:
    """Serialize the request sent from the ROS 2 bridge to the Julia server.

    `stop=True` asks the server to finish and exit; it replies once with an empty
    control set. Used so a client can shut the brain down without a signal, which
    keeps Ctrl-C semantics out of the normal teardown path.
    """
    return json.dumps(
        {
            "version": PROTOCOL_VERSION,
            "t": float(t),
            "stop": bool(stop),
            "agents": [
                {
                    "id": int(a.id),
                    "kind": str(a.kind),
                    "p": [float(x) for x in a.p],
                    "v": [float(x) for x in a.v],
                    "yaw": float(a.yaw),
                    "ok": bool(a.ok),
                }
                for a in agents
            ],
        }
    ).encode()


def parse_control(raw: bytes | str) -> ControlOutput:
    """Deserialize the reply from the Julia server."""
    d = json.loads(raw)
    version = d.get("version", PROTOCOL_VERSION)
    if version != PROTOCOL_VERSION:
        raise ValueError(f"protocol mismatch: got v{version}, expected v{PROTOCOL_VERSION}")
    return ControlOutput(
        agents=[
            AgentControl(
                id=int(a["id"]),
                u=[float(x) for x in a["u"]],
                yaw_rate=float(a.get("yaw_rate", 0.0)),
            )
            for a in d["agents"]
        ],
        targets=[TargetPose(id=int(t["id"]), p=[float(x) for x in t["p"]]) for t in d.get("targets", [])],
        diag=dict(d.get("diag", {})),
    )

File: examples/ros2/protocol.jl
# protocol.jl — Julia side of the sheaf/ROS 2 wire protocol (v2).
#
# Mirrors examples/ros2/ros2_ws/src/sheaf_ros2/sheaf_ros2/protocol.py EXACTLY.
# Change the message format here AND in protocol.py together — nowhere else.
#
# v2 vs the SheafSim v1 protocol (SheafSim/julia/ground/protocol.jl):
#   v1 is unicycle-specific: state (x, y, psi, v, omega), control (v, omega).
#   v2 is dynamics-agnostic: state is a 3-D pose/twist, control is a world-frame
#   ENU velocity. The conversion to (v, omega) for a differential-drive base, or
#   to a PX4 velocity setpoint, happens in the ROS 2 node on the robot — the
#   sheaf controller never learns about wheels.
#
# Frames: everything is ENU, in the mocap / Gazebo world frame. Metres, rad, s.

using JSON3

const PROTOCOL_VERSION = 2

struct AgentState
    id   :: Int
    kind :: Symbol            # :px4 | :diffdrive
    p    :: Vector{Float64}   # position   [x, y, z]  (ENU, world)
    v    :: Vector{Float64}   # velocity   [vx, vy, vz]
    yaw  :: Float64           # heading about +z
    ok   :: Bool              # false => state is stale/invalid, hold this agent
end

struct SystemState
    t      :: Float64
    agents :: Vector{AgentState}
    stop   :: Bool            # client is going away; shut the server down cleanly
end

SystemState(t, agents) = SystemState(t, agents, false)

struct AgentControl
    id       :: Int
    u        :: Vector{Float64}   # commanded world-frame ENU velocity [ux, uy, uz]
    yaw_rate :: Float64
end

struct TargetPose
    id :: Int
    p  :: Vector{Float64}
end

struct ControlOutput
    agents  :: Vector{AgentControl}
    targets :: Vector{TargetPose}
    diag    :: Dict{String, Any}
end

ControlOutput(agents, targets) = ControlOutput(agents, targets, Dict{String,Any}())

function parse_state(raw::AbstractString)::SystemState
    d = JSON3.read(raw)
    haskey(d, :version) && d.version != PROTOCOL_VERSION &&
        error("protocol mismatch: got v$(d.version), expected v$PROTOCOL_VERSION")
    agents = [AgentState(
        a.id,
        Symbol(a.kind),
        Float64[a.p[1], a.p[2], a.p[3]],
        Float64[a.v[1], a.v[2], a.v[3]],
        Float64(a.yaw),
        Bool(a.ok),
    ) for a in d.agents]
    return SystemState(Float64(d.t), agents, get(d, :stop, false))
end

function format_control(ctrl::ControlOutput)::String
    JSON3.write(Dict(
        "version" => PROTOCOL_VERSION,
        "agents"  => [Dict(
            "id"       => a.id,
            "u"        => a.u,
            "yaw_rate" => a.yaw_rate,
        ) for a in ctrl.agents],
        "targets" => [Dict("id" => t.id, "p" => t.p) for t in ctrl.targets],
        "diag"    => ctrl.diag,
    ))
end
